File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/LocationService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.dto.LocationCreationDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.LocationDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.LocationFilterDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.LocationForLenderSearchDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.LocationTagCollectionDto;
import at.ac.tuwien.sepm.groupphase.backend.dto.PageableDto;
import at.ac.tuwien.sepm.groupphase.backend.enums.LocationSortingCriterion;

public interface LocationService {

    /**
     * Finds all locations that contain the given string in either their name or their owner's name,
     * sorted by the given criterion.
     *
     * @param name     String to search for
     * @param page     Page to request
     * @param pageSize Size of the page to request
     * @param sortingCriterion Criterion to sort results by
     * @return Found locations as pageable
     */
    PageableDto<LocationDto> searchByName(String name, int page, int pageSize, LocationSortingCriterion sortingCriterion);

    /**
     * Finds all locations that match the given Filter-params,
     * sorted by the given criterion.
     *
     * @param locationFilterDto     Params to search for
     * @param page     Page to request
     * @param pageSize Size of the page to request
     * @param sortingCriterion Criterion to sort results by
     * @return Found locations as pageable
     */
    PageableDto<LocationDto> filter(LocationFilterDto locationFilterDto, int page, int pageSize, LocationSortingCriterion sortingCriterion);

    /**
     * Finds all locations that are owned by a lender with the given id,
     * sorted by a given criterion.
     *
     * @param searchDto Includes id of lender, paging information,
     *                  flag whether the include removed locations in the result
     *                  and a sorting criterion for the results.
     * @return  Found locations as pageable
     */
    PageableDto<LocationDto> searchByLender(LocationForLenderSearchDto searchDto);


    /**
     * Creates a new location.
     *
     * @param locationDto Location to create
     * @param user        User creating the location
     * @return Created location with id
     */
    LocationDto createLocation(LocationCreationDto locationDto, String user);

    /**
     * Returns the location with the given id.
     *
     * @param id ID of the location to return
     * @param user User requesting the location
     * @return LocationDto Location with the given id
     */
    LocationDto getLocationById(Long id, String user);

    /**
     * Updates the location with the given id (only the owner can update the location).
     *
     * @param id ID of the location to update
     * @param locationDto Location to update
     * @param user User updating the location
     * @return LocationDto Location with the given id
     */
    LocationDto updateLocation(Long id, String user, LocationDto locationDto);

    /**
     * Removes the location from the system (only hides location to the outside).
     * Action can only be performed by the owner.
     *
     * @param id ID of the location to remove
     * @param user User updating the location
     * @return LocationDto The removed location
     */
    LocationDto removeLocation(Long id, String user);

    /**
     * Removes all locations that are not deleted and that belong to the given user.
     *
     * @param email email address of given user
     */
    void removeAllExistingLocationsForLender(String email);

    /**
     * Returns all tags that are currently existing in the database.
     *
     * @return LocationTagCollectionDto Collection of all tags
     */
    LocationTagCollectionDto getAllTags();
}

File: frontend/src/app/services/location.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Globals } from 'src/app/global';
import { LocationDto, Pageable, LocationTagCollectionDto, LocationFilterDto } from 'src/app/dtos';
import { LocationSortingCriterion } from 'src/app/enum';

@Injectable({
  providedIn: 'root',
})
export class LocationService {
  private locationBaseUri: string = this.globals.backendUri + '/locations';
  private locationSearchUri: string = this.locationBaseUri + '/search';

  constructor(private httpClient: HttpClient, private globals: Globals) {}

  /**
   * Searches for locations based on if the given query string is contained within
   * a location's or it's lender's name.
   *
   * The result is given as page of all results from the query, sorted by the given
   * or the default criterion.
   *
   * @param q The query string
   * @param page The page to request (starts at 0)
   * @param pageSize The size of the requested page
   * @param sort Criterion to sort locations by (default: recommended)
   * @returns A wrapper for the found location with page information
   */
  search(
    q: string,
    page: number,
    pageSize: number,
    sort: LocationSortingCriterion = LocationSortingCriterion.RECOMMENDED_DESC
  ): Observable<Pageable<LocationDto>> {
    console.log(`Search for locations with q=${q}, page=${page}, pageSize=${pageSize}`);
    return this.httpClient.get<Pageable<LocationDto>>(`${this.locationBaseUri}/search`, {
      params: { q, page, pageSize, sort },
    });
  }

  /**
   * Filters for locations based on the given Filter Information.
   *
   * The result is given as page of all results from the query, sorted by the given
   * or the default criterion.
   *
   * @param filter The query string
   * @param page The page to request (starts at 0)
   * @param pageSize The size of the requested page
   * @param sort Criterion to sort locations by (default: recommended)
   * @returns A wrapper for the found location with page information
   */
  filter(
    filterDto: LocationFilterDto,
    page: number,
    pageSize: number,
    sort: LocationSortingCriterion = LocationSortingCriterion.RECOMMENDED_DESC
  ): Observable<Pageable<LocationDto>> {
    console.log(filterDto);
    console.log(`Filter locations with filter(
      searchString=${filterDto.searchString},
      plz=${filterDto.plz},
      state=${filterDto.state},
      address=${filterDto.address},
      timeFrom=${filterDto.timeFrom},
      timeTo=${filterDto.timeTo},
      priceFrom=${filterDto.priceFrom},
      priceTo=${filterDto.priceTo}),
      page=${page}, pageSize=${pageSize}`);
    return this.httpClient.post<Pageable<LocationDto>>(this.locationSearchUri, filterDto, {
      params: { page, pageSize, sort },
    });
  }

  /**
   * Searches for locations based on the id of the lender of the location, orders
   * results by the given or default criterion.
   *
   * @param id The id of the lender of the location
   * @param includeRemovedLocations Flag whether to include removed locations in the result
   * @param page The page to request (first is 0)
   * @param pageSize The size of the requested page
   * @param sort Criterion to sort locations by (default: recommended)
   * @returns A wrapper for the found location with page information
   */
  searchByLender(
    id: number,
    includeRemovedLocations: boolean,
    page: number,
    pageSize: number,
    sort: LocationSortingCriterion = LocationSortingCriterion.RECOMMENDED_DESC
  ): Observable<Pageable<LocationDto>> {
    return this.httpClient.get<Pageable<LocationDto>>(`${this.locationBaseUri}/search/lender`, {
      params: { id, page, includeRemovedLocations, pageSize, sort },
    });
  }

  /**
   * Creates a new location with the given data.
   *
   * @param location The location to create
   * @returns The created location (with id)
   */
  create(location: LocationDto): Observable<LocationDto> {
    return this.httpClient.post<LocationDto>(`${this.locationBaseUri}`, location);
  }

  /**
   * Updates the location with the given id with the given data.
   *
   * @param id The id of the location to update
   * @param location The location data to update
   * @returns The updated location
   */
  update(id: number, location: LocationDto): Observable<LocationDto> {
    return this.httpClient.put<LocationDto>(`${this.locationBaseUri}/${id}`, location);
  }

  /**
   * Returns the location with the given id.
   *
   * @param id The id of the location
   * @returns The location with the given id
   */
  getById(id: number): Observable<LocationDto> {
    console.log(`Get location with id=${id}`);
    return this.httpClient.get<LocationDto>(`${this.locationBaseUri}/${id}`);
  }

  /**
   * Removes the location with the given id.
   *
   * @param id The id of the location
   * @returns The removed location
   */
  remove(id: number): Observable<LocationDto> {
    console.log(`Remove location with id=${id}`);
    return this.httpClient.delete<LocationDto>(`${this.locationBaseUri}/${id}`);
  }

  /**
   * Returns all tags that currently exist.
   *
   * @returns All tags that currently exist
   */
  getAllTags(): Observable<LocationTagCollectionDto> {
    return this.httpClient.get<LocationTagCollectionDto>(`${this.locationBaseUri}/tags`);
  }
}
